### ISSUE TITLE
Glide fails to load thumbnails after rotation

### ISSUE THREAD

**Issue by x0b** [x0b/rcloneExplorer#34]
Glide fails to load images after a device rotation.

**Stacktrace**
```
2019-12-26 18:10:01.319 16366-16366/ca.pkay.rcloneexplorer.x0b W/Glide: Load failed for http://127.0.0.1:29179/<redacted-access-key>/example.PNG with size [120x120]
    class com.bumptech.glide.load.engine.GlideException: Failed to load resource
    There was 1 cause:
    java.net.ConnectException(Connection refused)
     call GlideException#logRootCauses(String) for more detail
      Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetching data failed, class java.io.InputStream, REMOTE
    There was 1 cause:
    java.net.ConnectException(Connection refused)
     call GlideException#logRootCauses(String) for more detail
        Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetch failed
    There was 1 cause:
    java.net.ConnectException(Connection refused)
     call GlideException#logRootCauses(String) for more detail
          Cause (1 of 1): class java.net.ConnectException: Connection refused
2019-12-26 18:10:01.321 16366-16366/ca.pkay.rcloneexplorer.x0b I/Glide: Root cause (1 of 1)
    java.net.ConnectException: Connection refused
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:343)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:205)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:187)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:356)
        at java.net.Socket.connect(Socket.java:600)
        at com.android.okhttp.internal.Platform.connectSocket(Platform.java:113)
        at com.android.okhttp.Connection.connectSocket(Connection.java:196)
        at com.android.okhttp.Connection.connect(Connection.java:172)
        at com.android.okhttp.Connection.connectAndSetOwner(Connection.java:367)
        at com.android.okhttp.OkHttpClient$1.connectAndSetOwner(OkHttpClient.java:130)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:330)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:247)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:457)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:126)
        at com.bumptech.glide.load.data.HttpUrlFetcher.loadDataWithRedirects(HttpUrlFetcher.java:104)
        at com.bumptech.glide.load.data.HttpUrlFetcher.loadData(HttpUrlFetcher.java:59)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.loadData(MultiModelLoader.java:100)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.startNextOrFail(MultiModelLoader.java:164)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.onLoadFailed(MultiModelLoader.java:154)
        at com.bumptech.glide.load.data.HttpUrlFetcher.loadData(HttpUrlFetcher.java:65)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.loadData(MultiModelLoader.java:100)
        at com.bumptech.glide.load.engine.SourceGenerator.startNext(SourceGenerator.java:62)
        at com.bumptech.glide.load.engine.DecodeJob.runGenerators(DecodeJob.java:309)
        at com.bumptech.glide.load.engine.DecodeJob.runWrapped(DecodeJob.java:279)
        at com.bumptech.glide.load.engine.DecodeJob.run(DecodeJob.java:235)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
        at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run(GlideExecutor.java:446)
```

**Error analysis so far**
- When rotating the device, the old thumbnail server is destroyed. Afterwards, Android re-initializes the view and then also restarts the thumbnail server with new parameters.
- In the current version, rclone sometimes can not bind to the fileserver port since it is still bound to the old instance
    - Fixed: not the root cause
 - Glide will try to load a new image preview from the thumbnail server before it is ready (race condition)
   - increasing the timeout does not work (the connection is refused immediately)
- the default http loader does not retry loading the image
- current ideas for solving this
   a. proxy the connection and stall until ready (really hacky)
   b. manipulate the caching keys so that glide uses the cached image (should be done anyway to reduce network traffic)
   c. make the service survice device rotations (probably best solution)

Update: option b works. I stabilized the caching key to a full qualified remote path,

**Comment by moarbytes** [comment]
Would adding embedded thumbnails to my media assist Glide with being able to (re)generate the thumbnail library? My knowledge of Android and the APIs it uses is still limited, but is it possible to replicate [this objective c method](https://developer.apple.com/documentation/avfoundation/avcapturephotosettings/2873905-embeddedthumbnailphotoformat?language=objc) in Glide?

**Comment by x0b** [comment]
Not really, the root cause of the problem is different and this would not solve it. The current problem is a consequence of the security mitigations introduced in version 1.9.0. 

Basically, since images are served with a random security key, the generated Glide cache id is always different and Glide will always experience a cache miss on rotation. Then, Glide will immediately try to load the image, but the rclone serve instance is only ready a few moments after and the load fails back to the fallback icon.

I'm currently testing this
> Update: option b works. I stabilized the caching key to a full qualified remote path

and it works fine so far because the Glide cache then works correctly. I'll include the fix in 1.10.1, but i just did not want to ship it in 1.10.0 without thorough testing.
